pkg/bucket/bandwidth/measurement.go

```diff
@@ -0,0 +1,133 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bandwidth
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// betaBucket is the weight used to calculate exponential moving average
+	betaBucket = 0.1 // Number of averages considered = 1/(1-betaObject)
+)
+
+// bucketMeasurement captures the bandwidth details for one bucket
+type bucketMeasurement struct {
+	lock                       sync.Mutex
+	cumulativeBytes            uint64        // Sum of all bytes transferred before the latest update
+	cumulativeTime             time.Duration // Total time taken for bytes
+	sumOfTimePerObject         time.Duration // Sum of time taken by objects in a window
+	sumOfTimePerObjectInWindow time.Duration // Sum of time taken by objects in a window
+	bytesSinceLastWindow       uint64        // Total bytes since last window was processed
+	windowStart                time.Time     // Start time for window
+	expMovingAvg               float64       // Previously calculate sliding window
+}
+
+// newBucketMeasurement creates a new instance of the measurement with the initial start time.
+func newBucketMeasurement(initTime time.Time) *bucketMeasurement {
+	return &bucketMeasurement{
+		windowStart: initTime,
+	}
+}
+
+// incrementBytes add bytes reported for a bucket.
+func (m *bucketMeasurement) incrementBytes(bytes uint64) {
+	atomic.AddUint64(&m.cumulativeBytes, bytes)
+	atomic.AddUint64(&m.bytesSinceLastWindow, bytes)
+}
+
+// calculateWindow processes the measurements between fixed time slots.
+func (m *bucketMeasurement) calculateWindow(windowEnd time.Time) {
+	// Calculate aggregate avg bandwidth and exp window avg
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	defer func() {
+		m.windowStart = windowEnd
+		m.sumOfTimePerObjectInWindow = 0
+	}()
+
+	if windowEnd.Before(m.windowStart) {
+		return
+	}
+
+	duration := windowEnd.Sub(m.windowStart)
+
+	// Do not update cumulative time if no traffic
+	bytesSinceLastWindow := atomic.SwapUint64(&m.bytesSinceLastWindow, 0)
+	if bytesSinceLastWindow > 0 {
+		m.cumulativeTime += duration
+		m.sumOfTimePerObject += m.sumOfTimePerObjectInWindow
+		// If total time spend copying object is less than wall clock time, then use
+		// sum of object to avoid  under reporting bandwidth.
```

Comment:
I don't understand. If we are measuring the bandwidth over the time of the window. The length should be the time of the window.

What problem are you trying to resolve?

Isn't the problem just that due to how you average you will only ever see 90% of the bandwidth right away?

Reply:
monitoring is reporting bandwidth per second but processes in a window of time. Let’s say only one object was transferred and it took half the window to transfer. The code before would divide the bytes with the size of the window which will underreport the aggregate bandwidth. The same is true for more than one object as long as sum of time taken per object is less than the window duration. But due to concurrency and traffic with lots of large files, when the sum is larger than wall clock window duration we should use the window duration... the change from before is to tune the bandwidth reported when there is low traffic and small files. This, change will improve the accuracy of reporting for bandwidth under light load and specifically for aggregate bandwidth.